contract/src/lib.rs

```diff
@@ -412,6 +426,45 @@ impl Contract {
             false
         }
     }
+
+    /// Public call to run GC on a mainchain.
+    /// batch_size is how many block headers should be removed in the execution
+    ///
+    /// # Panics
+    /// If initial blockheader or tip blockheader are not in a header pool
+    pub fn run_mainchain_gc(&mut self, batch_size: u64) {
+        let initial_blockheader = self
+            .headers_pool
+            .get(&self.mainchain_initial_blockhash)
+            .expect("initial blockheader must be in a header pool");
+
+        let tip_blockheader = self
+            .headers_pool
+            .get(&self.mainchain_tip_blockhash)
+            .expect("tip blockheader must be in a header pool");
+
+        let amount_of_headers_we_store =
+            tip_blockheader.block_height - initial_blockheader.block_height;
+
+        if amount_of_headers_we_store > self.gc_threshold {
+            let total_amount_to_remove = amount_of_headers_we_store - self.gc_threshold;
+            let selected_amount_to_remove = std::cmp::min(total_amount_to_remove, batch_size);
+
+            let start_removal_height = initial_blockheader.block_height;
+            let end_removal_height = initial_blockheader.block_height + selected_amount_to_remove;
+
+            for height in start_removal_height..end_removal_height {
+                let blockhash = self.mainchain_height_to_header[&height].clone();
+
+                self.mainchain_header_to_height.remove(&blockhash);
+                self.mainchain_height_to_header.remove(&height);
+                self.headers_pool.remove(&blockhash);
```

Comment:
```suggestion
                let blockhash = &self.mainchain_height_to_header[&height];

                self.headers_pool.remove(blockhash);
                self.mainchain_header_to_height.remove(blockhash);
                self.mainchain_height_to_header.remove(&height);
```

Reply:
Updated in https://github.com/aurora-is-near/btc-light-client-contract/commit/2c765163713994bb6cdd22daf20eba487a1b2c68